os/kernel/sched/sched_waitid.c

```diff
@@ -180,26 +180,25 @@ int waitid(idtype_t idtype, id_t id, FAR siginfo_t *info, int options)
 	int err;
 	int ret;
 
-	/* waitid() is a cancellation point */
-	(void)enter_cancellation_point();
-
 	/* MISSING LOGIC:   If WNOHANG is provided in the options, then this function
 	 * should returned immediately.  However, there is no mechanism available now
 	 * know if the thread has child:  The children remember their parents (if
 	 * CONFIG_SCHED_HAVE_PARENT) but the parents do not remember their children.
 	 */
 
-	/* None of the options are supported except for WEXITED (which must be
+	/* Verify idtype and options.
+	 * We support P_ALL and P_PID ID types only.
+	 * And none of the options are supported except for WEXITED (which must be
 	 * provided.  Currently SIGCHILD always reports CLD_EXITED so we cannot
 	 * distinguish any other events.
 	 */
-
-#ifdef CONFIG_DEBUG
-	if (options != WEXITED) {
+	if (((idtype != P_ALL) && (idtype != P_PID)) || (options != WEXITED)) {
 		set_errno(ENOSYS);
```

Comment:
Is it correct? I think EINVAL should be set to errno.
In posix documentation,
EINVAL  returns when an invalid value was specified for options, or idtype and id specify an invalid set of processes.